### PR TITLE
An alternative approach of updating the direction of the history term 

### PR DESCRIPTION
**Pull request by dhairyaiitb**
**Summary**

As mentioned in Luding (2008), the tangential displacement history in models needs to be updated to account for the rotation of the particle pair reference frame due to changes in the relative positions of the particles. The current approach removes any component of the net tangential displacement that lies along the current normal direction and rescales it to preserve its magnitude. However, this method results in unphysical particle sticking for larger size ratios. Therefore, we propose an alternate approach to project the history term directly in the direction of current tangential velocuty. This was also discussed during the online meeting with Prof. Paul Umbanhowar and Prof. Richard Lueptow.

**Related Issue(s)**

**Author(s)**

Dhairya R. Vyas, Postdoctoral researcher, Department of Mechanical Engineering, Northwestern University 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

The implementation is currently hard-coded. In `src/GRANULAR/gran_sub_mod_tangential.cpp`, I added a flag named "newapproach" to toggle between the newer and older models. It would be beneficial to enable this option in the LAMMPS input script, but I'm unsure how to implement that.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: examples/granular/history_test/particles.dat [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+1 2 0.0005 2500 0.005 2.7937481665024794E-05 0.0016005372526493155 0 0 0
+2 1 0.01 2500 0.01 0.0 0.0 0 0 0
+3 1 0.01 2500 0.0 0.0 0.0 0 0 0
+
+Velocities
+
+1 0  0 0 0 0 0
+2 0  0 0 0 0 0
+3 0  0 0 0 0 0
```
Comment (akohlmey) [inline examples/granular/history_test/particles.dat]: Please do not add files that are missing the final newline character.
Also, if this is a test case, this should not be added to the LAMMPS distribution but rather you make a .tar.gz archive from those files and attach that to the pull request instead.

File: examples/granular/history_test/seive.lammps [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+dump            1 all custom ${wfreq} op.dump id x y z vx vy vz type diameter 
+dump_modify     1 pad 10
+
+comm_modify     vel yes
+
+thermo          ${wfreq}
+thermo_style    custom step ke
+run_style       verlet
+run             ${Nsteps}
```
Comment (akohlmey) [inline examples/granular/history_test/seive.lammps]: Same as before: no final new line character

File: src/GRANULAR/gran_sub_mod_tangential.cpp [hunk 3]
```diff
@@ -123,16 +123,29 @@ void GranSubModTangentialLinearHistory::calculate_forces()
 
   // rotate and update displacements / force.
   // see e.g. eq. 17 of Luding, Gran. Matter 2008, v10,p235
+
+  // int newapproach = 0; //DRV: Original Method results in unphysical pendular motion
+  int newapproach = 1; //DRV: New Method
+
   if (gm->history_update) {
-    rsht = dot3(history, gm->nx);
+    if(newapproach == 0){
```
Comment (akohlmey) [inline src/GRANULAR/gran_sub_mod_tangential.cpp]: The LAMMPS programming style has a blank between "if" and "(", also ")" and "{"

File: src/GRANULAR/gran_sub_mod_tangential.cpp [hunk 4]
```diff
@@ -123,16 +123,29 @@ void GranSubModTangentialLinearHistory::calculate_forces()
...
+      rsht = dot3(history, gm->nx); 
+    }
```
Comment (akohlmey) [inline src/GRANULAR/gran_sub_mod_tangential.cpp]: The LAMMPS programming style puts the "else" on the same line as the closing brace "}" and there is a space between "else" and "{"

File: src/GRANULAR/gran_sub_mod_tangential.cpp [hunk 5]
```diff
@@ -123,16 +123,29 @@ void GranSubModTangentialLinearHistory::calculate_forces()
 
   // rotate and update displacements / force.
   // see e.g. eq. 17 of Luding, Gran. Matter 2008, v10,p235
+
+  // int newapproach = 0; //DRV: Original Method results in unphysical pendular motion
```
Comment (akohlmey) [inline src/GRANULAR/gran_sub_mod_tangential.cpp:128]: For compile time conditional code, please use C style preprocessor directives.

**Review by akohlmey — CHANGES_REQUESTED** [review]
@dhairyaiitb thanks for your submission. I've requested some cosmetic changes to have your code be consistent with the LAMMPS programming style.

**Comment by jtclemm** [comment]
Thank you Dhairya for the addition. Is there a reason you only made this change to a single tangential model? Why not Mindlin, for instance?

I can help with adding an input option to toggle methods, but I think @dsbolin is the expert who will need to weigh in on whether the numerics are appropriate.

**Comment by dhairyaiitb** [comment]
Oh yes, @jtclemm, I was previously using an older version and forgot to incorporate it into the newer one. Now, the Mindlin model also includes the updated approach. We are also studying the rolling term and working on a manuscript where we have performed several validation studies by comparing it with an analytical solution. The test case is now attached as a zip file as suggested by @akohlmey. 

[test_case.zip](https://github.com/user-attachments/files/16401141/test_case.zip)

File: examples/granular/history_test/particles.dat (hunk 1 above)
Comment (dhairyaiitb) [reply examples/granular/history_test/particles.dat]: Thanks! I've removed the test case from the examples and attached it as a zip file.

File: examples/granular/history_test/seive.lammps (hunk 2 above)
Comment (dhairyaiitb) [reply examples/granular/history_test/seive.lammps]: Thanks! I've removed the test case from the examples and attached it as a zip file.

File: src/GRANULAR/gran_sub_mod_tangential.cpp (hunk 3 above)
Comment (dhairyaiitb) [reply src/GRANULAR/gran_sub_mod_tangential.cpp]: Thanks, have incorporated these changes.

File: src/GRANULAR/gran_sub_mod_tangential.cpp (hunk 4 above)
Comment (dhairyaiitb) [reply src/GRANULAR/gran_sub_mod_tangential.cpp]: Thanks, have incorporated these changes.

File: src/GRANULAR/gran_sub_mod_tangential.cpp (hunk 5 above)
Comment (dhairyaiitb) [reply src/GRANULAR/gran_sub_mod_tangential.cpp:128]: This is temporary and will be removed in the next version, so I have not modified it for now.

**Comment by dhairyaiitb** [comment]
Hi, are there any updates on this? We can send an initial draft of the manuscript via email. The draft includes the validation studies and the potential consequences of not implementing this. Can you please share your email address?

**Comment by sjplimp** [comment]
@dhairyaiitb Joel's email is on the LAMMPS authors webpage

**Comment by dhairyaiitb** [comment]
Thank you. I will mail it to his email address.

**Comment by dhairyaiitb** [comment]
I am closing this pull request and will open another one with the revised version. Thanks.

**Pull request closed without merging**